File: platform/sdk/dotnet/MirrorSource.cs
#nullable enable

using System.Diagnostics;
using System.Globalization;
using System.Security.Cryptography;
using System.Text;
using System.Text.Json;

namespace LayerX.Sdk;

public sealed record MirrorCandidate(int Source, byte[] Commitment);
public sealed record MirrorPolicy(MirrorPolicyKind Kind, IReadOnlyList<MirrorCandidate> Candidates, int Minimum = 1);
public sealed record MirrorVerification(string Level, ulong BatchNumber, byte[] HeaderDigest,
    byte[] EvidenceDigest, string SourceId, string Target, string CanonicalPosition,
    string Provenance, ulong? LatestBatch, string BatchLag, int FailoverCount,
    int AgreeingSources, string CheckpointLevel);
public sealed class MirrorVerificationException(string code) : Exception($"mirror verification refused: {code}")
{
    public string Code { get; } = code;
}

public sealed class MirrorSourceVerifier
{
    private const int MaximumRequestBytes = 40 * 1024 * 1024;
    private const int MaximumResponseBytes = 1024 * 1024;
    private const int MaximumEvidenceBytes = (MaximumRequestBytes - 64 * 1024) / 2;
    private const long MaximumExecutableBytes = 512L * 1024 * 1024;
    private const long MaximumConfigurationBytes = 16L * 1024 * 1024;
    private static readonly HashSet<string> ErrorCodes = new(StringComparer.Ordinal)
    {
        "configuration", "unavailable", "rate-limited", "missing", "target-mismatch",
        "source-mismatch", "malformed", "bounds", "commitment", "authorization",
        "proof", "checkpoint-unavailable", "divergent", "insufficient-agreement", "reorged"
    };

    private readonly string executable;
    private readonly string configuration;
    private readonly byte[] executableDigest;
    private readonly byte[] configurationDigest;
    private readonly TimeSpan timeout;

    public MirrorSourceVerifier(string executable, string configuration, TimeSpan timeout)
    {
        if (timeout < TimeSpan.FromMilliseconds(100) || timeout > TimeSpan.FromSeconds(120))
        {
            throw new MirrorVerificationException("configuration");
        }
        var executableInput = TrustedInput(executable, true, MaximumExecutableBytes);
        var configurationInput = TrustedInput(configuration, false, MaximumConfigurationBytes);
        this.executable = executableInput.Path;
        this.configuration = configurationInput.Path;
        this.executableDigest = executableInput.Digest;
        this.configurationDigest = configurationInput.Digest;
        this.timeout = timeout;
    }

    public Task<MirrorVerification> VerifyReceiptAsync(ulong batch, MirrorPolicy policy,
        ReadOnlyMemory<byte> receipt, CancellationToken cancellation = default)
    {
        if (receipt.Length > MaximumEvidenceBytes) throw new MirrorVerificationException("bounds");
        return VerifyAsync(batch, policy, new Dictionary<string, object>
        {
            { "kind", "receipt" },
            { "canonical_hex", Convert.ToHexString(receipt.Span).ToLowerInvariant() }
        }, cancellation);
    }

    public Task<MirrorVerification> VerifyStateAsync(ulong batch, MirrorPolicy policy,
        ReadOnlyMemory<byte> state, ReadOnlyMemory<byte> proof,
        CancellationToken cancellation = default)
    {
        if (state.Length > MaximumEvidenceBytes || proof.Length > MaximumEvidenceBytes - state.Length)
        {
            throw new MirrorVerificationException("bounds");
        }
        return VerifyAsync(batch, policy, new Dictionary<string, object>
        {
            { "kind", "state" },
            { "canonical_hex", Convert.ToHexString(state.Span).ToLowerInvariant() },
            { "proof_hex", Convert.ToHexString(proof.Span).ToLowerInvariant() }
        }, cancellation);
    }

    private async Task<MirrorVerification> VerifyAsync(ulong batch, MirrorPolicy policy,
        Dictionary<string, object> evidence, CancellationToken cancellation)
    {
        if (batch == 0 || policy.Candidates.Count == 0
            || policy.Candidates.Count > MirrorSchemaV2.MaximumSources)
        {
            throw new MirrorVerificationException("configuration");
        }
        var seen = new HashSet<int>();
        var candidates = new List<object>();
        foreach (var value in policy.Candidates)
        {
            if (value.Source < 0 || !seen.Add(value.Source) || value.Commitment.Length != 32)
            {
                throw new MirrorVerificationException("configuration");
            }
            candidates.Add(new
            {
                source = value.Source,
                commitment_hex = Convert.ToHexString(value.Commitment).ToLowerInvariant()
            });
        }
        object wirePolicy = policy.Kind switch
        {
            MirrorPolicyKind.Exact when candidates.Count == 1 =>
                new { kind = "exact", candidate = candidates[0] },
            MirrorPolicyKind.OrderedPreference => new { kind = "ordered-preference", candidates },
            MirrorPolicyKind.Agreement when policy.Minimum > 0 && policy.Minimum <= candidates.Count =>
                new { kind = "agreement", candidates, minimum = policy.Minimum },
            _ => throw new MirrorVerificationException("configuration")
        };
        var request = JsonSerializer.SerializeToUtf8Bytes(new
        {
            batch_number = batch.ToString(CultureInfo.InvariantCulture),
            evidence,
            policy = wirePolicy
        });
        if (request.Length > MaximumRequestBytes) throw new MirrorVerificationException("bounds");

        RequireTrustedInputs();
        var start = new ProcessStartInfo(executable)
        {
            RedirectStandardInput = true,
            RedirectStandardOutput = true,
            RedirectStandardError = true,
            UseShellExecute = false
        };
        start.ArgumentList.Add(configuration);
        using var process = new Process { StartInfo = start };
        using var deadline = CancellationTokenSource.CreateLinkedTokenSource(cancellation);
        deadline.CancelAfter(timeout);
        try
        {
            if (!process.Start()) throw new MirrorVerificationException("unavailable");
            var outputTask = ReadBoundedAsync(process.StandardOutput.BaseStream,
                MaximumResponseBytes, deadline.Token);
            var errorTask = DrainAsync(process.StandardError.BaseStream, deadline.Token);
            var inputTask = WriteInputAsync(process, request, deadline.Token);
            await Task.WhenAll(inputTask, process.WaitForExitAsync(deadline.Token));
            var output = await outputTask;
            await errorTask;
            RequireTrustedInputs();
            if (process.ExitCode != 0) throw new MirrorVerificationException("unavailable");
            if (output.Exceeded) throw new MirrorVerificationException("bounds");
            return Parse(output.Bytes, batch, policy);
        }
        catch (OperationCanceledException)
        {
            if (!process.HasExited) process.Kill(true);
            throw new MirrorVerificationException("unavailable");
        }
        catch (IOException)
        {
            if (!process.HasExited) process.Kill(true);
            throw new MirrorVerificationException("unavailable");
        }
        catch (JsonException)
        {
            throw new MirrorVerificationException("malformed");
        }
        catch (InvalidOperationException)
        {
            throw new MirrorVerificationException("malformed");
        }
        catch (KeyNotFoundException)
        {
            throw new MirrorVerificationException("malformed");
        }
        catch (FormatException)
        {
            throw new MirrorVerificationException("malformed");
        }
    }

    private static async Task WriteInputAsync(Process process, byte[] request,
        CancellationToken cancellation)
    {
        await process.StandardInput.BaseStream.WriteAsync(request, cancellation);
        process.StandardInput.Close();
    }

    private static async Task<BoundedOutput> ReadBoundedAsync(Stream stream, int maximum,
        CancellationToken cancellation)
    {
        using var retained = new MemoryStream(Math.Min(maximum, 4096));
        var buffer = new byte[8192];
        var exceeded = false;
        int read;
        while ((read = await stream.ReadAsync(buffer, cancellation)) != 0)
        {
            var remaining = maximum - checked((int)retained.Length);
            var keep = Math.Min(Math.Max(remaining, 0), read);
            if (keep != 0) retained.Write(buffer, 0, keep);
            if (keep != read) exceeded = true;
        }
        return new BoundedOutput(retained.ToArray(), exceeded);
    }

    private static async Task DrainAsync(Stream stream, CancellationToken cancellation)
    {
        var buffer = new byte[8192];
        while (await stream.ReadAsync(buffer, cancellation) != 0) { }
    }

    private static MirrorVerification Parse(byte[] output, ulong requestedBatch,
        MirrorPolicy policy)
    {
        using var document = JsonDocument.Parse(output, new JsonDocumentOptions
        {
            MaxDepth = MirrorSchemaV2.MaximumJsonDepth
        });
        var root = document.RootElement;
        if (!root.TryGetProperty("ok", out var ok) || ok.ValueKind != JsonValueKind.True)
        {
            var code = root.TryGetProperty("error", out var error) && error.ValueKind == JsonValueKind.String
                ? error.GetString() ?? "malformed" : "malformed";
            throw new MirrorVerificationException(ErrorCodes.Contains(code) ? code : "malformed");
        }
        var value = root.GetProperty("verification");
        var batch = Unsigned(value.GetProperty("batchNumber"));
        var level = Text(value, "level", 64);
        var source = Text(value, "sourceId", 64);
        var target = Text(value, "target", 2048);
        var position = Text(value, "canonicalPosition", 2048);
        var provenance = Text(value, "provenance", 16);
        var lag = Text(value, "batchLag", 64);
        var checkpoint = Text(value, "checkpointLevel", 32);
        var failover = value.GetProperty("failoverCount").GetInt32();
        var agreeing = value.GetProperty("agreeingSources").GetInt32();
        if (batch != requestedBatch || (provenance != "Canonical" && provenance != "Reorged")
            || checkpoint != "unavailable" || failover < 0 || failover >= policy.Candidates.Count
            || agreeing < 1 || agreeing > policy.Candidates.Count
            || (policy.Kind == MirrorPolicyKind.Agreement && agreeing < policy.Minimum))
        {
            throw new MirrorVerificationException("malformed");
        }
        ulong? latestBatch = null;
        if (value.TryGetProperty("latestBatch", out var latest) && latest.ValueKind != JsonValueKind.Null)
        {
            latestBatch = Unsigned(latest);
        }
        return new MirrorVerification(level, batch, Digest(value, "headerDigest"),
            Digest(value, "evidenceDigest"), source, target, position, provenance,
            latestBatch, lag, failover, agreeing, checkpoint);
    }

    private sealed record TrustedMirrorInput(string Path, byte[] Digest);

    private static TrustedMirrorInput TrustedInput(string path, bool executable, long maximum)
    {
        try
        {
            if (!Path.IsPathFullyQualified(path)
                || !string.Equals(Path.GetFullPath(path), path, StringComparison.Ordinal))
            {
                throw new MirrorVerificationException("configuration");
            }
            for (var current = new FileInfo(path).Directory; current is not null; current = current.Parent)
            {
                if (!current.Exists || current.LinkTarget is not null
                    || (current.Attributes & FileAttributes.ReparsePoint) != 0)
                {
                    throw new MirrorVerificationException("configuration");
                }
                RequireProtectedMode(current.FullName);
            }
            var info = new FileInfo(path);
            if (!info.Exists || info.LinkTarget is not null
                || (info.Attributes & (FileAttributes.Directory | FileAttributes.ReparsePoint)) != 0
                || info.Length < 0 || info.Length > maximum)
            {
                throw new MirrorVerificationException("configuration");
            }
            RequireProtectedMode(path);
            if (executable && !OperatingSystem.IsWindows()
                && (File.GetUnixFileMode(path)
                    & (UnixFileMode.UserExecute | UnixFileMode.GroupExecute | UnixFileMode.OtherExecute)) == 0)
            {
                throw new MirrorVerificationException("configuration");
            }
            using var source = new FileStream(path, FileMode.Open, FileAccess.Read, FileShare.Read,
                64 * 1024, FileOptions.SequentialScan);
            using var hash = IncrementalHash.CreateHash(HashAlgorithmName.SHA256);
            var buffer = new byte[64 * 1024];
            long total = 0;
            int count;
            while ((count = source.Read(buffer)) != 0)
            {
                total += count;
                if (total > maximum) throw new MirrorVerificationException("configuration");
                hash.AppendData(buffer, 0, count);
            }
            info.Refresh();
            if (!info.Exists || total != info.Length) {
                throw new MirrorVerificationException("configuration");
            }
            return new TrustedMirrorInput(path, hash.GetHashAndReset());
        }
        catch (IOException)
        {
            throw new MirrorVerificationException("configuration");
        }
        catch (UnauthorizedAccessException)
        {
            throw new MirrorVerificationException("configuration");
        }
    }

    private static void RequireProtectedMode(string path)
    {
        if (OperatingSystem.IsWindows()) return;
        var unsafeModes = UnixFileMode.GroupWrite | UnixFileMode.OtherWrite;
        if ((File.GetUnixFileMode(path) & unsafeModes) != 0)
        {
            throw new MirrorVerificationException("configuration");
        }
    }

    private void RequireTrustedInputs()
    {
        var currentExecutable = TrustedInput(executable, true, MaximumExecutableBytes);
        var currentConfiguration = TrustedInput(configuration, false, MaximumConfigurationBytes);
        if (!CryptographicOperations.FixedTimeEquals(executableDigest, currentExecutable.Digest)
            || !CryptographicOperations.FixedTimeEquals(configurationDigest, currentConfiguration.Digest))
        {
            throw new MirrorVerificationException("configuration");
        }
    }

    private static string Text(JsonElement value, string name, int maximum)
    {
        var result = value.GetProperty(name).GetString();
        if (string.IsNullOrEmpty(result) || Encoding.UTF8.GetByteCount(result) > maximum)
        {
            throw new MirrorVerificationException("malformed");
        }
        return result;
    }

    private static ulong Unsigned(JsonElement value)
    {
        if (value.ValueKind != JsonValueKind.String) throw new MirrorVerificationException("malformed");
        var text = value.GetString();
        if (string.IsNullOrEmpty(text) || text[0] == '0'
            || !ulong.TryParse(text, NumberStyles.None, CultureInfo.InvariantCulture, out var result)
            || result == 0 || result.ToString(CultureInfo.InvariantCulture) != text)
        {
            throw new MirrorVerificationException("malformed");
        }
        return result;
    }

    private static byte[] Digest(JsonElement value, string name)
    {
        try
        {
            var result = Convert.FromHexString(Text(value, name, 64));
            if (result.Length != 32) throw new MirrorVerificationException("malformed");
            return result;
        }
        catch (FormatException)
        {
            throw new MirrorVerificationException("malformed");
        }
    }

    private sealed record BoundedOutput(byte[] Bytes, bool Exceeded);
}

File: platform/sdk/swift/Sources/LayerXSDK/MirrorSource.swift
import Foundation
import Crypto
#if canImport(Darwin)
import Darwin
#elseif canImport(Glibc)
import Glibc
#endif

public struct MirrorCandidate: Sendable {
    public let source: Int
    public let commitment: Data
    public init(source: Int, commitment: Data) {
        self.source = source
        self.commitment = commitment
    }
}

public struct MirrorPolicy: Sendable {
    public let kind: MirrorPolicyKind
    public let candidates: [MirrorCandidate]
    public let minimum: Int
    public init(kind: MirrorPolicyKind, candidates: [MirrorCandidate], minimum: Int = 1) {
        self.kind = kind
        self.candidates = candidates
        self.minimum = minimum
    }
}

public struct MirrorVerification: Sendable {
    public let level: String
    public let batchNumber: UInt64
    public let headerDigest: Data
    public let evidenceDigest: Data
    public let sourceID: String
    public let target: String
    public let canonicalPosition: String
    public let provenance: String
    public let latestBatch: UInt64?
    public let batchLag: String
    public let failoverCount: Int
    public let agreeingSources: Int
    public let checkpointLevel: String
}

public protocol MirrorSourceVerifying: Sendable {
    func receipt(batchNumber: UInt64, policy: MirrorPolicy,
                 canonicalReceipt: Data) throws -> MirrorVerification
    func state(batchNumber: UInt64, policy: MirrorPolicy, canonicalState: Data,
               canonicalProof: Data) throws -> MirrorVerification
}

#if os(macOS) || os(Linux)
public final class LocalMirrorExecutableVerifier: MirrorSourceVerifying, @unchecked Sendable {
    private static let maximumRequestBytes = 40 * 1024 * 1024
    private static let maximumResponseBytes = 1024 * 1024
    private static let maximumEvidenceBytes = (maximumRequestBytes - 64 * 1024) / 2
    private static let maximumExecutableBytes: UInt64 = 512 * 1024 * 1024
    private static let maximumConfigurationBytes: UInt64 = 16 * 1024 * 1024

    private let executable: URL
    private let configuration: URL
    private let executableDigest: Data
    private let configurationDigest: Data
    private let timeout: TimeInterval

    public init(executable: URL, configuration: URL, timeout: TimeInterval = 120) throws {
        guard timeout >= 0.1, timeout <= 120 else {
            throw MirrorErrorCode.configuration
        }
        let executableInput = try Self.trustedInput(
            executable, executable: true, maximum: Self.maximumExecutableBytes)
        let configurationInput = try Self.trustedInput(
            configuration, executable: false, maximum: Self.maximumConfigurationBytes)
        self.executable = executableInput.url
        self.configuration = configurationInput.url
        self.executableDigest = executableInput.digest
        self.configurationDigest = configurationInput.digest
        self.timeout = timeout
    }

    public func receipt(batchNumber: UInt64, policy: MirrorPolicy,
                        canonicalReceipt: Data) throws -> MirrorVerification {
        guard canonicalReceipt.count <= Self.maximumEvidenceBytes else {
            throw MirrorErrorCode.bounds
        }
        return try verify(batchNumber: batchNumber, policy: policy,
                          evidence: ["kind": "receipt", "canonical_hex": canonicalReceipt.hex])
    }

    public func state(batchNumber: UInt64, policy: MirrorPolicy, canonicalState: Data,
                      canonicalProof: Data) throws -> MirrorVerification {
        guard canonicalState.count <= Self.maximumEvidenceBytes,
              canonicalProof.count <= Self.maximumEvidenceBytes - canonicalState.count else {
            throw MirrorErrorCode.bounds
        }
        return try verify(batchNumber: batchNumber, policy: policy,
                          evidence: ["kind": "state", "canonical_hex": canonicalState.hex,
                                     "proof_hex": canonicalProof.hex])
    }

    private func verify(batchNumber: UInt64, policy: MirrorPolicy,
                        evidence: [String: Any]) throws -> MirrorVerification {
        guard batchNumber > 0, !policy.candidates.isEmpty,
              policy.candidates.count <= MirrorSchemaV2.maximumSources else {
            throw MirrorErrorCode.configuration
        }
        var seen = Set<Int>()
        let candidates = try policy.candidates.map { value -> [String: Any] in
            guard value.source >= 0, seen.insert(value.source).inserted,
                  value.commitment.count == 32 else {
                throw MirrorErrorCode.configuration
            }
            return ["source": value.source, "commitment_hex": value.commitment.hex]
        }
        let wirePolicy: [String: Any]
        switch policy.kind {
        case .exact:
            guard candidates.count == 1 else { throw MirrorErrorCode.configuration }
            wirePolicy = ["kind": "exact", "candidate": candidates[0]]
        case .orderedPreference:
            wirePolicy = ["kind": "ordered-preference", "candidates": candidates]
        case .agreement:
            guard policy.minimum > 0, policy.minimum <= candidates.count else {
                throw MirrorErrorCode.configuration
            }
            wirePolicy = ["kind": "agreement", "candidates": candidates,
                          "minimum": policy.minimum]
        }
        let request = try JSONSerialization.data(withJSONObject:
            ["batch_number": batchNumber.description, "evidence": evidence,
             "policy": wirePolicy])
        guard request.count <= Self.maximumRequestBytes else { throw MirrorErrorCode.bounds }

        let process = Process()
        let input = Pipe()
        let output = Pipe()
        try requireTrustedInputs()
        process.executableURL = executable
        process.arguments = [configuration.path]
        process.standardInput = input
        process.standardOutput = output
        process.standardError = FileHandle.nullDevice
        let retained = BoundedOutput(maximum: Self.maximumResponseBytes)
        let inputState = ProcessInputState()
        let inputGroup = DispatchGroup()
        let outputGroup = DispatchGroup()
        do {
            try process.run()
        } catch {
            throw MirrorErrorCode.unavailable
        }
        outputGroup.enter()
        DispatchQueue.global(qos: .utility).async {
            defer { outputGroup.leave() }
            do {
                while let chunk = try output.fileHandleForReading.read(upToCount: 8192),
                      !chunk.isEmpty {
                    retained.append(chunk)
                }
            } catch {
                retained.fail()
            }
        }
        inputGroup.enter()
        DispatchQueue.global(qos: .utility).async {
            defer {
                try? input.fileHandleForWriting.close()
                inputGroup.leave()
            }
            do {
                try input.fileHandleForWriting.write(contentsOf: request)
            } catch {
                inputState.fail()
            }
        }

        let deadline = DispatchTime.now() + .milliseconds(Int(timeout * 1_000))
        while process.isRunning && DispatchTime.now() < deadline {
            Thread.sleep(forTimeInterval: 0.005)
        }
        if process.isRunning {
            process.terminate()
            let grace = DispatchTime.now() + .milliseconds(100)
            while process.isRunning && DispatchTime.now() < grace {
                Thread.sleep(forTimeInterval: 0.005)
            }
            if process.isRunning {
                Self.forceKill(process.processIdentifier)
            }
            try? input.fileHandleForWriting.close()
            try? output.fileHandleForReading.close()
            throw MirrorErrorCode.unavailable
        }
        guard inputGroup.wait(timeout: .now() + .seconds(1)) == .success,
              outputGroup.wait(timeout: .now() + .seconds(1)) == .success,
              !inputState.failed, !retained.failed, process.terminationStatus == 0 else {
            throw MirrorErrorCode.unavailable
        }
        try requireTrustedInputs()
        guard !retained.exceeded else { throw MirrorErrorCode.bounds }
        return try parse(retained.data, requestedBatch: batchNumber, policy: policy)
    }

    private func parse(_ bytes: Data, requestedBatch: UInt64,
                       policy: MirrorPolicy) throws -> MirrorVerification {
        guard let response = try JSONSerialization.jsonObject(with: bytes) as? [String: Any]
        else { throw MirrorErrorCode.malformed }
        guard response["ok"] as? Bool == true else {
            throw MirrorErrorCode(rawValue: response["error"] as? String ?? "malformed")
                ?? MirrorErrorCode.malformed
        }
        guard let value = response["verification"] as? [String: Any],
              let level = text(value["level"], maximum: 64),
              let batch = uint64(value["batchNumber"]), batch == requestedBatch,
              let header = Data(hex: value["headerDigest"] as? String), header.count == 32,
              let digest = Data(hex: value["evidenceDigest"] as? String), digest.count == 32,
              let source = text(value["sourceId"], maximum: 64),
              let target = text(value["target"], maximum: 2048),
              let position = text(value["canonicalPosition"], maximum: 2048),
              let provenance = text(value["provenance"], maximum: 16),
              provenance == "Canonical" || provenance == "Reorged",
              let lag = text(value["batchLag"], maximum: 64),
              let failover = integer(value["failoverCount"]), failover >= 0,
              failover < policy.candidates.count,
              let agreeing = integer(value["agreeingSources"]), agreeing > 0,
              agreeing <= policy.candidates.count,
              policy.kind != .agreement || agreeing >= policy.minimum,
              let checkpoint = text(value["checkpointLevel"], maximum: 32),
              checkpoint == "unavailable" else {
            throw MirrorErrorCode.malformed
        }
        let latest: UInt64?
        if let raw = value["latestBatch"], !(raw is NSNull) {
            guard let parsed = uint64(raw) else { throw MirrorErrorCode.malformed }
            latest = parsed
        } else {
            latest = nil
        }
        return MirrorVerification(level: level, batchNumber: batch, headerDigest: header,
            evidenceDigest: digest, sourceID: source, target: target,
            canonicalPosition: position, provenance: provenance, latestBatch: latest,
            batchLag: lag, failoverCount: failover, agreeingSources: agreeing,
            checkpointLevel: checkpoint)
    }

    private struct TrustedInput {
        let url: URL
        let digest: Data
    }

    private static func trustedInput(_ url: URL, executable: Bool,
                                     maximum: UInt64) throws -> TrustedInput {
        let normalized = url.standardizedFileURL
        guard url.isFileURL, url.path.hasPrefix("/"), normalized.path == url.path,
              url.resolvingSymlinksInPath().path == url.path else {
            throw MirrorErrorCode.configuration
        }
        var current = url
        while current.path != "/" {
            let values = try current.resourceValues(forKeys: [
                .isRegularFileKey, .isDirectoryKey, .isSymbolicLinkKey,
            ])
            guard values.isSymbolicLink != true else { throw MirrorErrorCode.configuration }
            if current.path == url.path {
                guard values.isRegularFile == true else { throw MirrorErrorCode.configuration }
            } else {
                guard values.isDirectory == true else { throw MirrorErrorCode.configuration }
            }
            try requireProtectedOwnerAndMode(current)
            current.deleteLastPathComponent()
        }
        try requireProtectedOwnerAndMode(current)
        guard !executable || FileManager.default.isExecutableFile(atPath: url.path) else {
            throw MirrorErrorCode.configuration
        }
        let before = try FileManager.default.attributesOfItem(atPath: url.path)
        guard let declaredSize = (before[.size] as? NSNumber)?.uint64Value,
              declaredSize <= maximum else { throw MirrorErrorCode.configuration }
        let handle = try FileHandle(forReadingFrom: url)
        defer { try? handle.close() }
        var hasher = SHA256()
        var total: UInt64 = 0
        while let chunk = try handle.read(upToCount: 64 * 1024), !chunk.isEmpty {
            total += UInt64(chunk.count)
            guard total <= maximum else { throw MirrorErrorCode.configuration }
            hasher.update(data: chunk)
        }
        let after = try FileManager.default.attributesOfItem(atPath: url.path)
        guard total == declaredSize,
              stableIdentity(before) == stableIdentity(after) else {
            throw MirrorErrorCode.configuration
        }
        return TrustedInput(url: url, digest: Data(hasher.finalize()))
    }

    private static func requireProtectedOwnerAndMode(_ url: URL) throws {
        let attributes = try FileManager.default.attributesOfItem(atPath: url.path)
        guard let owner = (attributes[.ownerAccountID] as? NSNumber)?.uint32Value,
              owner == 0 || owner == geteuid(),
              let permissions = (attributes[.posixPermissions] as? NSNumber)?.uint16Value,
              permissions & 0o022 == 0 else {
            throw MirrorErrorCode.configuration
        }
    }

    private static func stableIdentity(_ attributes: [FileAttributeKey: Any]) -> String {
        let system = (attributes[.systemNumber] as? NSNumber)?.stringValue ?? ""
        let file = (attributes[.systemFileNumber] as? NSNumber)?.stringValue ?? ""
        let size = (attributes[.size] as? NSNumber)?.stringValue ?? ""
        let modified = (attributes[.modificationDate] as? Date)?.timeIntervalSince1970.description ?? ""
        return "\(system):\(file):\(size):\(modified)"
    }

    private func requireTrustedInputs() throws {
        let currentExecutable = try Self.trustedInput(
            executable, executable: true, maximum: Self.maximumExecutableBytes)
        let currentConfiguration = try Self.trustedInput(
            configuration, executable: false, maximum: Self.maximumConfigurationBytes)
        guard currentExecutable.digest == executableDigest,
              currentConfiguration.digest == configurationDigest else {
            throw MirrorErrorCode.configuration
        }
    }

    private static func forceKill(_ identifier: Int32) {
        #if canImport(Darwin)
        _ = Darwin.kill(identifier, SIGKILL)
        #elseif canImport(Glibc)
        _ = Glibc.kill(identifier, SIGKILL)
        #endif
    }
}

private final class BoundedOutput: @unchecked Sendable {
    private let lock = NSLock()
    private let maximum: Int
    private var retained = Data()
    private(set) var exceeded = false
    private(set) var failed = false

    init(maximum: Int) { self.maximum = maximum }

    func append(_ chunk: Data) {
        lock.lock()
        defer { lock.unlock() }
        let keep = min(max(maximum - retained.count, 0), chunk.count)
        if keep > 0 { retained.append(contentsOf: chunk.prefix(keep)) }
        if keep != chunk.count { exceeded = true }
    }

    func fail() {
        lock.lock()
        failed = true
        lock.unlock()
    }

    var data: Data {
        lock.lock()
        defer { lock.unlock() }
        return retained
    }
}

private final class ProcessInputState: @unchecked Sendable {
    private let lock = NSLock()
    private var value = false
    func fail() {
        lock.lock()
        value = true
        lock.unlock()
    }
    var failed: Bool {
        lock.lock()
        defer { lock.unlock() }
        return value
    }
}

private extension Data {
    var hex: String { map { String(format: "%02x", $0) }.joined() }
    init?(hex: String?) {
        guard let hex, hex.count % 2 == 0 else { return nil }
        var value = Data()
        var index = hex.startIndex
        while index < hex.endIndex {
            let end = hex.index(index, offsetBy: 2)
            guard let byte = UInt8(hex[index..<end], radix: 16) else { return nil }
            value.append(byte)
            index = end
        }
        self = value
    }
}

private func uint64(_ value: Any?) -> UInt64? {
    guard let text = value as? String, !text.isEmpty, text.first != "0",
          text.allSatisfy({ $0.isASCII && $0.isNumber }), let result = UInt64(text),
          result > 0 else { return nil }
    return result
}

private func integer(_ value: Any?) -> Int? {
    guard let number = value as? NSNumber, !(value is Bool),
          let result = Int(number.stringValue), String(result) == number.stringValue
    else { return nil }
    return result
}

private func text(_ value: Any?, maximum: Int) -> String? {
    guard let value = value as? String, !value.isEmpty, value.utf8.count <= maximum
    else { return nil }
    return value
}
#endif
